8293562: KeepAliveCache Blocks Threads while Closing Connections

PoojaDP-23 opened the pull request:
fix: https://bugs.openjdk.org/browse/JDK-8293562

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293562](https://bugs.openjdk.org/browse/JDK-8293562): KeepAliveCache Blocks Threads while Closing Connections (**Bug** - P4)


### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev.git pull/1890/head:pull/1890` \
`$ git checkout pull/1890`

Update a local copy of the PR: \
`$ git checkout pull/1890` \
`$ git pull https://git.openjdk.org/jdk11u-dev.git pull/1890/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1890`

View PR using the GUI difftool: \
`$ git pr show -t 1890`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1890.diff">https://git.openjdk.org/jdk11u-dev/pull/1890.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk11u-dev/pull/1890#issuecomment-1557112259)

jerboaa commented:
Seems a duplicate of #1825

alograg commented:
Hi @PoojaDP-23,
I was working on the backporting of the fix. #1825 
But after the @jerboaa reviews, I realized that your code is closer to the original fix commit than the mine.
I offer you all my support, I remain at your disposal.

jerboaa commented:
@PoojaDP-23 Please use `Backport 03f25a9c6924430ec4063b801b2b6ca55b9067c9` as the PR title so that bots recognize it as a backport. Also, please enable Github Action testing on your fork.

PoojaDP-23 commented:
> Hi @PoojaDP-23, I was working on the backporting of the fix. #1825 But after the @jerboaa reviews, I realized that your code is closer to the original fix commit than the mine. I offer you all my support, I remain at your disposal.

Thanks

jerboaa commented:
@PoojaDP-23 Please also merge latest `master` tree into the PR branch so as to trigger GHA checks for this PR. Thanks!

PoojaDP-23 commented:
@jerboaa - Yes, I have enabled  Github Action testing on my fork

PoojaDP-23 commented:
@jerboaa - I have performed below steps to merge latest master into PR branch. Could you please take a look?
```
git checkout master
git pull https://github.com/openjdk/jdk17u-dev master
git checkout <Your_BRANCH_NAME>
git merge master
git push origin HEAD

alograg commented:
Hi @jerboaa 
Can you do the review ?
Thrnaks,

jerboaa commented on src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java:
```diff
@@ -268,8 +290,8 @@ private void readObject(ObjectInputStream stream)
 
-/* FILO order for recycling HttpClients, should run in a thread
- * to time them out.  If > maxConns are in use, block.
+ /* LIFO order for reusing HttpClients. Most recent entries at the front.
```
Nit: Extraneous space in front of `/*`.

jerboaa commented on test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java:
```diff
@@ -0,0 +1,240 @@
+                            "Wait for second request was interrupted"));
+                }
+            } else {
+                System.out.println("name of the thread" +threadName);
```
This isn't part of the original change. Please remove.

jerboaa commented on test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java:
```diff
@@ -0,0 +1,240 @@
+                        "Close invoked from unexpected thread"));
+            }
+            System.out.println("Connection closed");
+            System.out.println(threadName);
```
Same here. Extraneous debug `println()`?

jerboaa reviewed (commented):
This seems OK to me. A few minor comments. What testing have you done?

darinhoward commented:
The docs at https://www.oracle.com/java/technologies/javase/11-0-17-relnotes.html indicate this fix was done in 11.0.17.0.3.  I'm assuming this is in error based on this PR still being open.

alograg commented:
Hi @darinhoward 
The bug is fixed in version JDK 11.0.17.0.3 (Oracle) not in [OpenJDK 11.0.17](https://bugs.openjdk.org/browse/JDK-8293562)

jerboaa commented:
> The docs at https://www.oracle.com/java/technologies/javase/11-0-17-relnotes.html indicate this fix was done in 11.0.17.0.3. I'm assuming this is in error based on this PR still being open.

Yes, those are Oracle JDK release notes. This PR is for OpenJDK.

PoojaDP-23 commented:
@jerboaa - I have addressed review comments and I have enabled GHA actions to test and I have separately run the testcase "B8293562.java" with(passes) and without(fails) the fix to test these changes. Thanks

PoojaDP-23 commented:
/integrate

PoojaDP-23 commented:
/label

PoojaDP-23 commented:
@jerboaa - Could you please add the fix request to the jbs issue? Please let me know if there are any actions from my side. Thanks

jerboaa commented:
> @jerboaa - Could you please add the fix request to the jbs issue? Please let me know if there are any actions from my side. Thanks

Done. Please mention what extra testing of the patch you've done here (other than the regtest). Thanks!

GoeLin commented:
Hi @PoojaDP-23 
You should not integrate before you have the fix-yes label!!!

I assume JDK-8296438, which is listed as related issue, is a follow up.

jerboaa commented:
Please check https://bugs.openjdk.org/browse/JDK-8296943 before this gets sponsored.

PoojaDP-23 commented:
@GoeLin - Apologies for commenting integrate. @jerboaa - I have enabled GHA actions and also I have ran the TIER1 and TIER2 tests and all are passing and I have ran the jtreg test [test/jdk/sun/net/www/http/KeepAliveCache/B8293562.java](https://github.com/openjdk/jdk11u-dev/pull/1890/files#diff-43af219708466772a71875d5d53dbfb340fe012b521788268716c242c668a834) which passes with this fix and fails without the fix
TIER: 1
Test summary
```
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:tier1                     1482  1482     0     0
   jtreg:test/jdk:tier1                               1881  1881     0     0
   jtreg:test/langtools:tier1                         3934  3934     0     0
   jtreg:test/nashorn:tier1                              0     0     0     0
   jtreg:test/jaxp:tier1                                 0     0     0     0
```

TEST SUCCESS


TIER: 2
Test summary

  ```
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:tier2                      325   325     0     0
   jtreg:test/jdk:tier2                               3468  3468     0     0
   jtreg:test/langtools:tier2                           11    11     0     0
   jtreg:test/nashorn:tier2                             36    36     0     0
   jtreg:test/jaxp:tier2                               440   440     0     0

```

TEST SUCCESS

PoojaDP-23 commented:
@jerboaa - I have checked https://bugs.openjdk.org/browse/JDK-8296943 and I have validated the test sun/net/www/http/HttpClient/MultiThreadTest.java with this fix and test is passing. Please take a look on below screenshot.

**Java-version:**
<img width="556" alt="image" src="https://github.com/openjdk/jdk11u-dev/assets/55285644/ed2f7857-55a3-4f76-8575-99efc9949552">




**Test-result**
<img width="944" alt="image" src="https://github.com/openjdk/jdk11u-dev/assets/55285644/8ff89095-eb57-4972-bbe8-e4d9d76f0c47">

jerboaa commented:
OK, thanks!

jerboaa commented:
/sponsor
